src/object.hpp

```diff
@@ -1172,31 +1231,102 @@ borrowed_object writable_object<Derived>::emplace(std::string_view key, T &&valu
     return emplace(key, owned_object{std::forward<T>(value)});
 }
 
-inline borrowed_object::borrowed_object(owned_object &obj) : obj_(obj.ptr()) {}
+inline borrowed_object::borrowed_object(owned_object &obj) : obj_(obj.ptr()), alloc_(obj.alloc()) {}
 
 // NOLINTNEXTLINE(cppcoreguidelines-rvalue-reference-param-not-moved)
 inline borrowed_object &borrowed_object::operator=(owned_object &&obj)
 {
-    ref() = obj.move();
+    alloc_ = obj.alloc();
+    *obj_ = obj.move();
     return *this;
 }
+namespace object_builder {
 
-inline owned_object owned_object::make_array(
-    std::initializer_list<detail::initializer::movable_object> list)
+struct movable_object {
+    // NOLINTNEXTLINE(google-explicit-constructor,hicpp-explicit-conversions)
+    template <typename T> movable_object(T &&value) : object{std::forward<T>(value)} {}
+    movable_object(const movable_object &) = delete;
+    movable_object(movable_object &&) = delete;
+    movable_object operator=(const movable_object &) = delete;
+    movable_object operator=(movable_object &&) = delete;
+    ~movable_object() = default;
+    mutable owned_object object;
```

Comment:
this hack would merit a comment explaining why it's needed

Reply:
Addressed in #420, in https://github.com/DataDog/libddwaf/pull/420/commits/bdbf8bbc892cd9e692ef61faf6f6382312fb78e1